README.md
# Scaleway Rubygem

[![Build Status](https://travis-ci.org/bchatelard/scaleway-ruby.svg?branch=develop)](https://travis-ci.org/bchatelard/scaleway-ruby)
[![Gem Version](https://badge.fury.io/rb/scaleway.svg)](http://badge.fury.io/rb/scaleway)

Easy to use Scaleway api client.

## Installation

Manual instalation:

```bash
gem install scaleway
```

add this to your gemfile:

```bash
gem 'scaleway'
```

## Usage

### Configure the client

```ruby
require 'scaleway'

Scaleway.organization = <organization_key>
Scaleway.token = <token>

# You can specify the zone, the default zone is 'par1'
Scaleway.zone = 'ams1'
```

### Servers

```ruby
# list all servers
Scaleway::Server.all

# list with filter
Scaleway::Server.all state: :running

# create a new server with default values
Scaleway::Server.create

# create a new server with name and tags
Scaleway::Server.create :name => 'my_new_server', tags: ['prod']

# get a server by id
server = Scaleway::Server.find <server_id>

# edit a server
server = Scaleway::Server.find <server_id>
server.name = 'new_name'
Scaleway::Server.edit server.id, server

# actions on a server
Scaleway::Server.power_on server.id

Scaleway::Server.power_off server.id

Scaleway::Server.terminate server.id

# destroy a server
Scaleway::Server.destroy server.id
```

### Marketplace

```ruby
# list all marketplace images
Scaleway::Marketplace.all

# get local image by id
image = Scaleway::Marketplace.find_local_image <marketplace_image_id>

# get local image by marketplace image id for arch arm
image = Scaleway::Marketplace.find_local_image <marketplace_image_id>, arch: 'arm'

# get local image by name
image = Scaleway::Marketplace.find_local_image_by_name("Docker")
```

### Images

```ruby
# list all images
Scaleway::Image.all

# get image by id
image = Scaleway::Image.find <image_id>

# get image by name
image = Scaleway::Image.find_by_name('Ubuntu')

# create an image
image = Scaleway::Image.create root_volume: snapshot

# edit an image
image = Scaleway::Image.edit id, image

# destroy an image
image = Scaleway::Image.destroy id
```

### Volumes

```ruby
# list all volumes
Scaleway::Volume.all

# get volume by id
volume = Scaleway::Volume.find <volume_id>

# create an volume
volume = Scaleway::Volume.create
volume = Scaleway::Volume.create size: 100 * 10 ** 9, volume_type: 'l_ssd'

# edit an volume
volume = Scaleway::Volume.edit id, volume

# destroy an volume
volume = Scaleway::Volume.destroy id
```

### Snapshots

```ruby
# list all snapshots
Scaleway::Volume.all

# get snapshot by id
snapshot = Scaleway::Snapshot.find <snapshot_id>

# create an snapshot
snapshot = Scaleway::Snapshot.create volume_id: id

# edit an snapshot
snapshot = Scaleway::Snapshot.edit id, snapshot

# destroy an snapshot
snapshot = Scaleway::Snapshot.destroy id
```

### Ip addresses

```ruby
# list all ip addresses
Scaleway::Ip.all

# get Ip
ip = Scaleway::Ip.find <ip_id>
ip = Scaleway::Ip.find 127.0.0.1

# reserve an ip
ip = Scaleway::Ip.reserve
# or
ip = Scaleway::Ip.create

# edit an ip
ip = Scaleway::Ip.edit id, ip

# release an ip
ip = Scaleway::Ip.destroy id
```

### Handle exceptions

```ruby
# Not found
begin
	puts Scaleway::Server.find <invalid_id>
rescue Scaleway::NotFound => e
	# handle error here
end

# Other
begin
	puts Scaleway::Server.create extra_field: ['nope']
rescue Scaleway::APIError => e
	# handle error here
end
```

## Example

```ruby
require 'scaleway'

Scaleway.organization = <organization_key>
Scaleway.token = <token>

# get the docker image from the marketplace
image = Scaleway::Marketplace.find_local_image_by_name('Docker')

# create 5 new servers
5.times do |x|
	Scaleway::Server.create name: "docker#{x}", image: image.id, tags: ['docker']
end

# power on
Scaleway::Server.all.each do |server|
	Scaleway::Server.power_on(server.id)
end

# do something ...

# terminate
Scaleway::Server.all.each do |server|
	Scaleway::Server.terminate(server.id)
end
```


CHANGES.md
# ChangeLog

## 1.0.1 (Unreleased)

- No entry.

## 1.0.0 (2016-10-31)

- Add explicit error on missing authentication.
- Add zone option.
- Update default commercial type to VC1S.
- Add marketplace API.

## 0.2.1 (2015-04-32)

- Fix unittests.

## 0.2.0 (2015-04-23)

- Rename to scaleway.
- Add the ability to reboot.

## 0.1.1 (2014-11-14)

- Fix push invalid gem.

## 0.1.0 (2014-11-14)

- Initial setup.


spec/servers_spec.rb
require 'rubygems'
require 'its'
require 'scaleway'

describe Scaleway::Server do
  subject(:server) { described_class }

  before do
    stubs = Faraday::Adapter::Test::Stubs.new do |stub|
      stub.get('/servers')      { |env| [200, {}, {:servers => [test_server]}] }
      stub.get('/servers/1234') { |env| [200, {}, {:server => test_server}] }
      stub.get('/servers/not_found') { |env| [404, {}, {:type => 'not_found', :message => 'not found'}] }
    end
    Scaleway.request = Faraday.new do |builder|
      builder.adapter :test, stubs
    end
  end

  describe "all" do
    let(:test_server)       { RecursiveOpenStruct.new({:name => 'hello', :id => '1234'}, :recurse_over_arrays => true) }

    its(:all)               { should eq [test_server] }
  end

  describe "item" do
    let(:test_server)       { RecursiveOpenStruct.new({:name => 'hello', :id => '1234'}, :recurse_over_arrays => true) }

    its(:find, '1234')      { should eq test_server }
  end

  describe "not found" do
    let(:test_server)       { RecursiveOpenStruct.new({:name => 'hello', :id => '1234'}, :recurse_over_arrays => true) }

    it { expect{Scaleway::Server.find('not_found')}.to raise_error(Scaleway::NotFound) }
  end

end


describe Scaleway::Server do
  subject(:server) { described_class }

  before do
    stubs = Faraday::Adapter::Test::Stubs.new do |stub|
      stub.get('/servers?state=running') { |env| [200, {}, {:servers => []}] }
      stub.get('/servers?invalid_filter=42') { |env| [400, {}, {:type => 'invalid_filter', :message => 'invalid filter'}] }
    end
    Scaleway.request = Faraday.new do |builder|
      builder.adapter :test, stubs
    end
  end

  describe "filter" do
    its(:all, state: 'running') { should eq [] }
  end

  describe "invalid filter" do
    it { expect{Scaleway::Server.all(invalid_filter: 42)}.to raise_error(Scaleway::APIError) }
  end
end


spec/scaleway_spec.rb
require 'rubygems'
require 'its'
require 'scaleway'

describe Scaleway do
  subject(:Scaleway) { described_class }

  before do
    Scaleway.organization = organization
    Scaleway.token = token
    Scaleway.zone = zone
  end

  describe "defaults" do
    let(:organization)   { nil }
    let(:token)          { nil }
    let(:zone)           { nil }

    its(:compute_endpoint) { should eq "https://cp-par1.scaleway.com" }
    its(:account_endpoint) { should eq "https://account.scaleway.com" }
    its(:token)            { should eq "token_required" }
    its(:organization)     { should eq "organization_required" }

    it { expect(Scaleway::VERSION).to eq "1.0.1" }
  end

  describe "test ams1" do
    let(:organization)   { nil }
    let(:token)          { nil }
    let(:zone)           { "ams1" }

    its(:compute_endpoint) { should eq "https://cp-ams1.scaleway.com" }
  end

  describe "test token and organization" do
    let(:organization)   { "organization_id" }
    let(:token)          { "token_id" }
    let(:zone)           { nil }

    its(:organization)   { should eq "organization_id" }
    its(:token)          { should eq "token_id" }
  end
end


spec/onlinelabs_spec.rb
require 'rubygems'
require 'its'
require 'onlinelabs'

describe OnlineLabs do
  subject(:OnlineLabs) { described_class }

  before do
    OnlineLabs.organization = organization
    OnlineLabs.token = token
  end

  describe "defaults" do
    let(:organization)   { nil }
    let(:token)          { nil }

    its(:compute_endpoint) { should eq "https://api.cloud.online.net" }
    its(:account_endpoint) { should eq "https://account.cloud.online.net" }
    its(:token)            { should eq "token_required" }
    its(:organization)     { should eq "organization_required" }

    it { expect(OnlineLabs::VERSION).to eq "0.1.2" }
  end

  describe "test token and organization" do
    let(:organization)   { "organization_id" }
    let(:token)          { "token_id" }

    its(:organization)   { should eq "organization_id" }
    its(:token)          { should eq "token_id" }
  end
end


lib/onlinelabs.rb
require 'faraday'
require 'faraday_middleware'
require 'recursive-open-struct'
require 'onlinelabs/version'

module OnlineLabs

  extend self

  def compute_endpoint
    "https://api.cloud.online.net"
  end

  def account_endpoint
    "https://account.cloud.online.net"
  end

  def token=(token)
    @token = token
    setup!
    @token
  end

  def token
    return @token if @token
    "token_required"
  end

  def organization=(organization)
    @organization = organization
    setup!
    @organization
  end

  def organization
    return @organization if @organization
    "organization_required"
  end

  DEFINITIONS = {
    "Image" => {
      :all => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images",
        :default_params => {
          :organization => Proc.new { OnlineLabs.organization }
        }
      },
      :marketplace => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images",
        :default_params => {
          :public => true
        }
      },
      :find => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images/%s",
      },
      :find_by_name => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images",
        :filters => [
          Proc.new { |item, params| item.name.include? params.first }
        ],
        :transform => Proc.new { |item, params| item.first },
      },
      :create => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images",
        :default_params => {
          :name => 'default',
          :root_volume => 'required',
          :organization => Proc.new { OnlineLabs.organization },
        }
      },
      :edit => {
        :method => :put,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images/%s",
      },
      :destroy => {
        :method => :delete,
        :endpoint => "#{OnlineLabs.compute_endpoint}/images/%s",
      },
    },
    "Volume" => {
      :all => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/volumes",
      },
      :find => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/volumes/%s",
      },
      :edit => {
        :method => :put,
        :endpoint => "#{OnlineLabs.compute_endpoint}/volumes/%s",
      },
      :destroy => {
        :method => :delete,
        :endpoint => "#{OnlineLabs.compute_endpoint}/volumes/%s",
      },
      :create => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/volumes",
        :default_params => {
          :name => 'default',
          :size => 20 * 10**9,
          :volume_type => 'l_hdd',
          :organization => Proc.new { OnlineLabs.organization },
        }
      },
    },
    "Ip" => {
      :all => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/ips",
      },
      :find => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/ips/%s",
      },
      :edit => {
        :method => :put,
        :endpoint => "#{OnlineLabs.compute_endpoint}/ips/%s",
      },
      :destroy => {
        :method => :delete,
        :endpoint => "#{OnlineLabs.compute_endpoint}/ips/%s",
      },
      :reserve => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/ips",
        :default_params => {
          :organization => Proc.new { OnlineLabs.organization },
        }
      },
      :create => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/ips",
        :default_params => {
          :organization => Proc.new { OnlineLabs.organization },
        }
      },
    },
    "Snapshot" => {
      :all => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/snapshots",
      },
      :find => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/snapshots/%s",
      },
      :edit => {
        :method => :put,
        :endpoint => "#{OnlineLabs.compute_endpoint}/snapshots/%s",
      },
      :destroy => {
        :method => :delete,
        :endpoint => "#{OnlineLabs.compute_endpoint}/snapshots/%s",
      },
      :create => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/snapshots",
        :default_params => {
          :name => 'default',
          :volume_id => 'required',
          :organization => Proc.new { OnlineLabs.organization },
        }
      },
    },
    "Server" => {
      :all => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers",
      },
      :power_on => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers/%s/action",
        :default_params => {
          :action => :poweron
        }
      },
      :power_off => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers/%s/action",
        :default_params => {
          :action => :poweroff
        }
      },
      :terminate => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers/%s/action",
        :default_params => {
          :action => :terminate
        }
      },
      :find => {
        :method => :get,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers/%s",
      },
      :edit => {
        :method => :put,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers/%s",
      },
      :destroy => {
        :method => :delete,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers/%s",
      },
      :create => {
        :method => :post,
        :endpoint => "#{OnlineLabs.compute_endpoint}/servers",
        :default_params => {
          :name => 'default',
          :image => Proc.new { OnlineLabs::Image.find_by_name('Ubuntu').id },
          :volumes => {},
          :organization => Proc.new { OnlineLabs.organization },
        }
      }
    },
  }

  DEFINITIONS.each do |resource|
    resource_name = resource[0]

    resource_class = Class.new(Object) do
      singleton = class << self; self end

      DEFINITIONS[resource_name].each do |method_name, query|
        singleton.send :define_method, method_name do |*args|
          OnlineLabs.request_and_respond(query, *args)
        end
      end
    end

    OnlineLabs.const_set(resource_name, resource_class)
  end

  def request=(request)
    @request = request
  end

  def request
    @request
  end

  def request_and_respond(query, *params)
    body = {}
    body.merge! query[:default_params] || {}
    endpoint = query[:endpoint]

    params.each do |param|
      if param.is_a? Hash or param.is_a? RecursiveOpenStruct
        body.merge! param
      elsif not param.nil?
        endpoint = endpoint % param
      end
    end

    body = Hash[body.map { |k, v|
      if v.respond_to? :call then [k, v.call()] else [k, v] end
    }]

    resp = OnlineLabs.request.send(query[:method], endpoint, body)
    body = resp.body
    if resp.status == 204
      return
    end

    if resp.status == 404
      raise OnlineLabs::NotFound, resp
    elsif resp.status >= 300
      raise OnlineLabs::APIError, resp
    end

    hash = RecursiveOpenStruct.new(body, :recurse_over_arrays => true)

    if body.length == 1
      hash = hash.send(body.keys.first)
    end

    if query[:filters]
      filters = query[:filters]
      hash = hash.find_all do |item|
        filters.all? do |filter|
          filter.call(item, params)
        end
      end
    end

    if query[:transform]
      hash = query[:transform].call(hash, params)
    end

    hash
  end

  class APIError < Exception
    def initialize(response)
      self.status = response.status
      self.body = RecursiveOpenStruct.new(response.body, :recurse_over_arrays => true)
      self.type = self.body.type
      self.error_message = self.body.message
    end

    def to_s
      "<status:#{status}, type:#{type}, message:\'#{error_message}\'>"
    end

    def message
      "#{self}"
    end

    attr_accessor :status
    attr_accessor :type
    attr_accessor :error_message
    attr_accessor :body
  end

  class NotFound < OnlineLabs::APIError
  end



  private

  def setup!
    options = {
      :headers => {
        'x-auth-token' => OnlineLabs.token,
        'content-type' => 'application/json',
      }
    }

    OnlineLabs.request = ::Faraday::Connection.new(options) do |builder|
      builder.use     ::FaradayMiddleware::EncodeJson
      builder.use     ::FaradayMiddleware::ParseJson
      builder.use     ::FaradayMiddleware::FollowRedirects
      builder.adapter ::Faraday.default_adapter
    end
  end
end


lib/scaleway.rb
require 'faraday'
require 'faraday_middleware'
require 'recursive-open-struct'
require 'scaleway/version'

module Scaleway

  extend self

  def compute_endpoint
    "https://cp-#{Scaleway.zone}.scaleway.com"
  end

  def account_endpoint
    "https://account.scaleway.com"
  end

  def marketplace_endpoint
    "https://api-marketplace.scaleway.com"
  end

  def zone=(zone)
    @zone = zone
    @zone
  end

  def zone
    return @zone || 'par1'
  end

  def token=(token)
    @token = token
    setup!
    @token
  end

  def token
    return @token if @token
    "token_required"
  end

  def organization=(organization)
    @organization = organization
    setup!
    @organization
  end

  def organization
    return @organization if @organization
    "organization_required"
  end

  DEFINITIONS = {
    "Marketplace" => {
      :all => {
        :method => :get,
        :endpoint => "#{Scaleway.marketplace_endpoint}/images",
      },
      :find_local_image_by_name => {
        :method => :get,
        :endpoint => "#{Scaleway.marketplace_endpoint}/images",
        :default_params => {
            arch: 'x86_64',
        },
        :filters => [
          Proc.new { |params, body, item|
            item.name.include? params.first }
        ],
        :transform => Proc.new { |params, body, item|
          item[0].versions[0].local_images.keep_if do |local_image|
            if local_image.zone == Scaleway.zone and local_image.arch == body[:arch]
              true
            end
          end.first
        },
      },
      :find_local_image => {
        :method => :get,
        :endpoint => "#{Scaleway.marketplace_endpoint}/images/%s/versions/current",
        :default_params => {
            arch: 'x86_64',
        },
        :transform => Proc.new { |params, body, item|
          item.local_images.keep_if do |local_image|
            if local_image.zone == Scaleway.zone and local_image.arch == body[:arch]
              true
            end
          end.first
        },
      },
    },
    "Image" => {
      :all => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/images" },
        :default_params => {
          :public => true
        }
      },
      :find => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/images/%s" },
      },
      :find_by_name => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/images" },
        :filters => [
          Proc.new { |params, body, item| item.name.include? params.first }
        ],
        :transform => Proc.new { |params, body, item| item.first },
      },
      :create => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/images" },
        :default_params => {
          :name => 'default',
          :root_volume => 'required',
          :organization => Proc.new { Scaleway.organization },
        }
      },
      :edit => {
        :method => :put,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/images/%s" },
      },
      :destroy => {
        :method => :delete,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/images/%s" },
      },
    },
    "Volume" => {
      :all => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/volumes" },
      },
      :find => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/volumes/%s" },
      },
      :edit => {
        :method => :put,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/volumes/%s" },
      },
      :destroy => {
        :method => :delete,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/volumes/%s" },
      },
      :create => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/volumes" },
        :default_params => {
          :name => 'default',
          :size => 20 * 10**9,
          :volume_type => 'l_hdd',
          :organization => Proc.new { Scaleway.organization },
        }
      },
    },
    "Ip" => {
      :all => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/ips" },
      },
      :find => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/ips/%s" },
      },
      :edit => {
        :method => :put,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/ips/%s" },
      },
      :destroy => {
        :method => :delete,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/ips/%s" },
      },
      :reserve => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/ips" },
        :default_params => {
          :organization => Proc.new { Scaleway.organization },
        }
      },
      :create => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/ips" },
        :default_params => {
          :organization => Proc.new { Scaleway.organization },
        }
      },
    },
    "Snapshot" => {
      :all => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/snapshots" },
      },
      :find => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/snapshots/%s" },
      },
      :edit => {
        :method => :put,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/snapshots/%s" },
      },
      :destroy => {
        :method => :delete,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/snapshots/%s" },
      },
      :create => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/snapshots" },
        :default_params => {
          :name => 'default',
          :volume_id => 'required',
          :organization => Proc.new { Scaleway.organization },
        }
      },
    },
    "Server" => {
      :all => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers" },
      },
      :power_on => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s/action" },
        :default_params => {
          :action => :poweron
        }
      },
      :reboot => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s/action" },
        :default_params => {
          :action => :reboot
        }
      },
      :power_off => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s/action" },
        :default_params => {
          :action => :poweroff
        }
      },
      :terminate => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s/action" },
        :default_params => {
          :action => :terminate
        }
      },
      :find => {
        :method => :get,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s" },
      },
      :edit => {
        :method => :put,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s" },
      },
      :destroy => {
        :method => :delete,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers/%s" },
      },
      :create => {
        :method => :post,
        :endpoint => Proc.new { "#{Scaleway.compute_endpoint}/servers" },
        :default_params => {
          :name => 'default',
          :commercial_type => 'VC1S',
          :image => Proc.new { |params, body|
            if body[:commercial_type] == 'C1'
              arch = 'arm'
            else
              arch = 'x86_64'
            end
            Scaleway::Marketplace.find_local_image_by_name(
              'Ubuntu Xenial', arch: arch).id
          },
          :volumes => {},
          :organization => Proc.new { Scaleway.organization },
        }
      }
    },
  }

  DEFINITIONS.each do |resource|
    resource_name = resource[0]

    resource_class = Class.new(Object) do
      singleton = class << self; self end

      DEFINITIONS[resource_name].each do |method_name, query|
        singleton.send :define_method, method_name do |*args|
          Scaleway.request_and_respond(query, *args)
        end
      end
    end

    Scaleway.const_set(resource_name, resource_class)
  end

  def request=(request)
    @request = request
  end

  def request
    @request
  end

  def request_and_respond(query, *params)
    body = {}
    body.merge! query[:default_params] || {}
    endpoint = query[:endpoint]

    if endpoint.respond_to? :call
      endpoint = endpoint.call()
    end

    params.each do |param|
      if param.is_a? Hash or param.is_a? RecursiveOpenStruct
        body.merge! param
      elsif not param.nil?
        endpoint = endpoint % param
      end
    end

    body = Hash[body.map { |k, v|
      if v.respond_to? :call then [k, v.call(params, body)] else [k, v] end
    }]

    if Scaleway.request.nil?
      raise ::RuntimeError.new("Authentication is missing")
    end

    resp = Scaleway.request.send(query[:method], endpoint, body)
    response_body = resp.body
    if resp.status == 204
      return
    end

    if resp.status == 404
      raise Scaleway::NotFound, resp
    elsif resp.status >= 300
      raise Scaleway::APIError, resp
    end

    hash = RecursiveOpenStruct.new(response_body, :recurse_over_arrays => true)

    if response_body.length == 1
      hash = hash.send(response_body.keys.first)
    end

    if query[:filters]
      filters = query[:filters]
      hash = hash.find_all do |item|
        filters.all? do |filter|
          filter.call(params, body, item)
        end
      end
    end

    if query[:transform]
      hash = query[:transform].call(params, body, hash)
    end

    hash
  end

  class APIError < Exception
    def initialize(response)
      self.status = response.status
      self.body = RecursiveOpenStruct.new(response.body, :recurse_over_arrays => true)
      self.type = self.body.type
      self.error_message = self.body.message
    end

    def to_s
      "<status:#{status}, type:#{type}, message:\'#{error_message}\'>"
    end

    def message
      "#{self}"
    end

    attr_accessor :status
    attr_accessor :type
    attr_accessor :error_message
    attr_accessor :body
  end

  class NotFound < Scaleway::APIError
  end



  private

  def setup!
    options = {
      :headers => {
        'x-auth-token' => Scaleway.token,
        'content-type' => 'application/json',
      }
    }

    Scaleway.request = ::Faraday::Connection.new(options) do |builder|
      builder.use     ::FaradayMiddleware::EncodeJson
      builder.use     ::FaradayMiddleware::ParseJson
      builder.use     ::FaradayMiddleware::FollowRedirects
      builder.adapter ::Faraday.default_adapter
    end
  end
end
